modules/apps/29-fee/ibc_module.go

```diff
@@ -171,32 +171,100 @@ func (im IBCModule) OnChanCloseConfirm(
 }
 
 // OnRecvPacket implements the IBCModule interface.
+// If fees are not enabled, this callback will default to the ibc-core packet callback
 func (im IBCModule) OnRecvPacket(
 	ctx sdk.Context,
 	packet channeltypes.Packet,
 	relayer sdk.AccAddress,
-) ibcexported.Acknowledgement {
-	// TODO: Implement fee specific logic if fee is enabled for the given channel
-	return im.app.OnRecvPacket(ctx, packet, relayer)
+) exported.Acknowledgement {
+	if !im.keeper.IsFeeEnabled(ctx, packet.SourcePort, packet.SourceChannel) {
+		return im.app.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	forwardRelayer, found := im.keeper.GetCounterpartyAddress(ctx, relayer.String())
+	if !found {
+		forwardRelayer = ""
+	}
+
+	return types.IncentivizedAcknowledgement{
+		Result:                ack.Acknowledgement(),
+		ForwardRelayerAddress: forwardRelayer,
+	}
 }
 
 // OnAcknowledgementPacket implements the IBCModule interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
 func (im IBCModule) OnAcknowledgementPacket(
 	ctx sdk.Context,
 	packet channeltypes.Packet,
 	acknowledgement []byte,
 	relayer sdk.AccAddress,
 ) error {
-	// TODO: Implement fee specific logic if fee is enabled for the given channel
-	return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if !im.keeper.IsFeeEnabled(ctx, packet.SourcePort, packet.SourceChannel) {
+		return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	ack := types.IncentivizedAcknowledgement{}
+	if err := types.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return sdkerrors.Wrapf(err, "cannot unmarshal ICS-29 incentivized packet acknowledgement: %v", ack)
+	}
+
+	packetId := channeltypes.NewPacketId(packet.SourceChannel, packet.SourcePort, packet.Sequence)
+	identifiedPacketFee, found := im.keeper.GetFeeInEscrow(ctx, packetId)
+
+	// return underlying callback if no fee found for given packetID
+	if !found {
+		return im.app.OnAcknowledgementPacket(ctx, packet, ack.Result, relayer)
+	}
+	// cache context before trying to distribute the fee
```

Comment:
```suggestion
	identifiedPacketFee, found := im.keeper.GetFeeInEscrow(ctx, packetId)
	if !found {
	        // return underlying callback if no fee found for given packetID
		return im.app.OnAcknowledgementPacket(ctx, packet, ack.Result, relayer)
	}
	
	// cache context before trying to distribute the fee
```

just a suggestion, might not want to commit via github UI since I don't think this is gofmt'd